File: src/todo/domain/events.py
"""Domain Events for the Todo bounded context.

Events are immutable records emitted by the ``Todo`` Aggregate *after* a
state change has been applied.  They describe what happened in the past and
carry enough data for downstream consumers to act without querying back.

Naming follows the Ubiquitous Language exactly:
  TodoCreated, TodoCompleted, TodoReopened, TodoTitleUpdated, TodoDeleted
"""
from __future__ import annotations

from dataclasses import dataclass

from todo.domain.value_objects import Timestamp, TodoId, TodoTitle


@dataclass(frozen=True)
class TodoCreated:
    """Emitted when a new Todo is successfully created."""

    todoId: TodoId
    title: TodoTitle
    occurredAt: Timestamp


@dataclass(frozen=True)
class TodoCompleted:
    """Emitted when a Todo transitions from ``active`` to ``completed``."""

    todoId: TodoId
    occurredAt: Timestamp


@dataclass(frozen=True)
class TodoReopened:
    """Emitted when a Todo transitions from ``completed`` back to ``active``."""

    todoId: TodoId
    occurredAt: Timestamp


@dataclass(frozen=True)
class TodoTitleUpdated:
    """Emitted when a Todo's title is changed."""

    todoId: TodoId
    newTitle: TodoTitle
    occurredAt: Timestamp


@dataclass(frozen=True)
class TodoDeleted:
    """Emitted when permanent removal of a Todo is requested.

    Actual removal is delegated to the ``TodoRepository``.
    """

    todoId: TodoId
    occurredAt: Timestamp

File: src/domain/events.ts
/**
 * Domain Events for the Todo bounded context.
 *
 * Events are immutable records emitted by the Todo Aggregate after a
 * state change has been applied. They describe what happened in the past.
 */

import type { TodoId, Timestamp, TodoTitle } from './value-objects';

// ---------------------------------------------------------------------------
// Individual event shapes (all fields readonly for immutability)
// ---------------------------------------------------------------------------

export interface TodoCreated {
  readonly type: 'TodoCreated';
  readonly todoId: TodoId;
  readonly title: TodoTitle;
  readonly occurredAt: Timestamp;
}

export interface TodoCompleted {
  readonly type: 'TodoCompleted';
  readonly todoId: TodoId;
  readonly occurredAt: Timestamp;
}

export interface TodoReopened {
  readonly type: 'TodoReopened';
  readonly todoId: TodoId;
  readonly occurredAt: Timestamp;
}

export interface TodoTitleUpdated {
  readonly type: 'TodoTitleUpdated';
  readonly todoId: TodoId;
  readonly newTitle: TodoTitle;
  readonly occurredAt: Timestamp;
}

export interface TodoDeleted {
  readonly type: 'TodoDeleted';
  readonly todoId: TodoId;
  readonly occurredAt: Timestamp;
}

// ---------------------------------------------------------------------------
// Discriminated union of all domain events
// ---------------------------------------------------------------------------

export type DomainEvent =
  | TodoCreated
  | TodoCompleted
  | TodoReopened
  | TodoTitleUpdated
  | TodoDeleted;
